### ISSUE TITLE
Performance Comparison Issue: Matrix Multiplication on MLX vs. PyTorch on Mac

### ISSUE THREAD

**Issue by Ysrae1** [ml-explore/mlx#1295]
### Environment
- **Mac Model:** MacBook Pro 14-inch, 2023
- **Processor:** M2 Pro
- **RAM:** 16GB
- **OS Version:** macOS 15.0
- **MLX Version:** 0.16.1  
- **PyTorch Version:** 2.4.0

### Issue Description
I conducted a performance test comparing matrix multiplication using MLX and PyTorch. I used a matrix of size 4000x4000 and ran tests 50 times to average the results. The performance difference is noticeable, and I seek to understand the underlying reasons and any possible optimizations.

### Test Code
```python
import mlx.core as mx
import torch
import time

size = 4000
num_run = 50

torch_time_cpu = 0
torch_time_gpu = 0
mlx_time_cpu = 0
mlx_time_gpu = 0

for i in range(num_run):

    a = torch.randn(size, size)
    b = torch.randn(size, size)
    a_gpu = torch.randn(size, size).to('mps')
    b_gpu = torch.randn(size, size).to('mps')
    c = mx.random.normal([size, size])
    d = mx.random.normal([size, size])

    start_time = time.time()
    result_torch = torch.matmul(a, b)
    torch_time_cpu_i = time.time() - start_time

    start_time = time.time()
    result_torch = torch.matmul(a_gpu, b_gpu)
    torch_time_gpu_i = time.time() - start_time

    start_time = time.time()
    mx.eval(mx.matmul(c, d, stream=mx.cpu))
    mlx_time_cpu_i = time.time() - start_time

    start_time = time.time()
    mx.eval(mx.matmul(c, d, stream=mx.gpu))
    mlx_time_gpu_i = time.time() - start_time

    torch_time_cpu += torch_time_cpu_i
    torch_time_gpu += torch_time_gpu_i
    mlx_time_cpu += mlx_time_cpu_i
    mlx_time_gpu += mlx_time_gpu_i

torch_time_cpu /= num_run
torch_time_gpu /= num_run
mlx_time_cpu /= num_run
mlx_time_gpu /= num_run

print(f'Matrix Multiplication Comparison. Size: [{size},{size}]x[{size},{size}]. Take the average of {num_run} trials.\n')

print(f"PyTorch CPU Time: {torch_time_cpu:.9f} seconds")

print(f"PyTorch GPU Time: {torch_time_gpu:.9f} seconds")

print(f"    MLX CPU Time: {mlx_time_cpu:.9f} seconds")

print(f"    MLX GPU Time: {mlx_time_gpu:.9f} seconds")

```

### Results

```python
Matrix Multiplication Comparison. Size: [4000,4000]x[4000,4000]. Take the average of 50 trials.

PyTorch CPU Time: 0.057260580 seconds
PyTorch GPU Time: 0.001417809 seconds
    MLX CPU Time: 0.112252812 seconds
    MLX GPU Time: 0.027218037 seconds
```

### Questions
1. Could you please help me understand the significant performance discrepancy on the GPU between MLX and PyTorch on the same MacBook Pro model?
2. Are there known advantages of using MLX in this scenario that might compensate for the observed performance gap?
3. Any suggestions or tips for optimizing MLX performance in matrix multiplication tasks would be greatly appreciated.

**Comment by angeloskath** [comment]
So the main problem in the above script is the way things are being measured. For PyTorch the MPS device needs to be synchronized before accessing the result and for MLX generally we would perform one evaluation for all 50 trials and not one evaluation per trial the same way that we would do a single synchronization for PyTorch.

Moreover it is customary to warmup the caches so that we can measure the performance of the operation and minimize the effect of the current system state. So I have edited the above script by reordering the tests a bit as follows:

- Run 5 warmup iterations before starting to measure
- Measure once for all `num_run` operations
- Run `b = a @ b` to make sure the matmuls depend on each other

On my M2 air I get the following results namely a ~10% speedup with MLX.

```
Matrix Multiplication Comparison. Size: [4000,4000]x[4000,4000]. Take the sum of 50 trials.

PyTorch CPU Time: 5.236518145 seconds
PyTorch GPU Time: 2.754643917 seconds
    MLX CPU Time: 5.221024036 seconds
    MLX GPU Time: 2.526622772 seconds
```

Here is the modified script

```python
import mlx.core as mx
import torch
import time

size = 4000
num_run = 50

torch_time_cpu = 0
torch_time_gpu = 0
mlx_time_cpu = 0
mlx_time_gpu = 0

print(f'Matrix Multiplication Comparison. Size: [{size},{size}]x[{size},{size}]. Take the sum of {num_run} trials.\n')

a = torch.randn(size, size)
b = torch.randn(size, size)
for i in range(5):
    b = torch.matmul(a, b)
start_time = time.time()
for i in range(num_run):
    b = torch.matmul(a, b)
torch_time_cpu = time.time() - start_time
print(f"PyTorch CPU Time: {torch_time_cpu:.9f} seconds")

a = torch.randn(size, size).to("mps")
b = torch.randn(size, size).to("mps")
for i in range(5):
    b = torch.matmul(a, b)
torch.mps.synchronize()
start_time = time.time()
for i in range(num_run):
    b = torch.matmul(a, b)
torch.mps.synchronize()
torch_time_gpu = time.time() - start_time
print(f"PyTorch GPU Time: {torch_time_gpu:.9f} seconds")

a = mx.random.normal(shape=(size, size))
b = mx.random.normal(shape=(size, size))
mx.eval(a, b)
with mx.stream(mx.cpu):
    for i in range(5):
        b = mx.matmul(a, b)
    mx.eval(b)
    start_time = time.time()
    for i in range(num_run):
        b = mx.matmul(a, b)
    mx.eval(b)
    mlx_time_cpu = time.time() - start_time
print(f"    MLX CPU Time: {mlx_time_cpu:.9f} seconds")

a = mx.random.normal(shape=(size, size))
b = mx.random.normal(shape=(size, size))
mx.eval(a, b)
with mx.stream(mx.gpu):
    for i in range(5):
        b = mx.matmul(a, b)
    mx.eval(b)
    start_time = time.time()
    for i in range(num_run):
        b = mx.matmul(a, b)
    mx.eval(b)
    mlx_time_gpu = time.time() - start_time
print(f"    MLX GPU Time: {mlx_time_gpu:.9f} seconds")
```

I am closing the issue but feel free to reopen it if you think that there is something missing from my answer.

**Comment by Ysrae1** [comment]
I have tried your code, but MLX still performs worse on GPU.

Here's my result.

```python
Matrix Multiplication Comparison. Size: [4000,4000]x[4000,4000]. Take the sum of 50 trials.

PyTorch CPU Time: 2.785820723 seconds
PyTorch GPU Time: 1.215188980 seconds
    MLX CPU Time: 2.660862923 seconds
    MLX GPU Time: 1.324872017 seconds
```

**Comment by angeloskath** [comment]
That is very plausible for just one size and device. We try to make sure that MLX is faster across many devices, operations and sizes. The community has done some extensive benchmarks for instance https://github.com/TristanBilot/mlx-benchmark/blob/main/benchmarks/average_benchmark.md which you can consult.

Furthermore I would benchmark the full application end to end since microbenchmarking a single matmul is not necessarily representative of the final performance.
